source/h1_stream.c

```diff
@@ -4,103 +4,174 @@
+        "Stream should be complete (or never-activated) when stream destroyed");
+    AWS_ASSERT(
+        aws_linked_list_empty(&stream->thread_data.chunk_list) &&
+        aws_linked_list_empty(&stream->synced_data.chunk_list) &&
+        "Chunks should be marked complete before stream destroyed");
 
-    aws_h1_stream_body_chunks_clean_up(stream);
     aws_h1_encoder_message_clean_up(&stream->encoder_message);
     aws_byte_buf_clean_up(&stream->incoming_storage_buf);
     aws_mem_release(stream->base.alloc, stream);
 }
 
-static void s_stream_update_window(struct aws_http_stream *stream, size_t increment_size) {
-    aws_http_connection_update_window(stream->owning_connection, increment_size);
+static struct aws_h1_connection *s_get_h1_connection(const struct aws_h1_stream *stream) {
+    return AWS_CONTAINER_OF(stream->base.owning_connection, struct aws_h1_connection, base);
+}
+
+static void s_stream_lock_synced_data(struct aws_h1_stream *stream) {
+    aws_h1_connection_lock_synced_data(s_get_h1_connection(stream));
 }
 
-static int s_body_chunks_init(struct aws_h1_stream *stream) {
-    AWS_PRECONDITION(stream);
-    aws_linked_list_init(&stream->body_chunks.chunk_list);
-    stream->body_chunks.current_chunk = NULL;
-    stream->body_chunks.paused = false;
-    return aws_mutex_init(&stream->body_chunks.lock);
+static void s_stream_unlock_synced_data(struct aws_h1_stream *stream) {
+    aws_h1_connection_unlock_synced_data(s_get_h1_connection(stream));
 }
 
-static void s_clean_up_body_chunk(struct aws_http1_stream_chunk **chunk) {
-    AWS_PRECONDITION(chunk);
-    aws_h1_stream_release_chunk(*chunk);
-    *chunk = NULL;
+static void s_stream_schedule_cross_thread_work_task(struct aws_h1_stream *stream) {
+    AWS_LOGF_TRACE(AWS_LS_HTTP_STREAM, "id=%p: Scheduling stream cross-thread work task.", (void *)&stream->base);
+
+    /* Keep stream alive until task completes */
+    aws_atomic_fetch_add(&stream->base.refcount, 1);
+    aws_channel_schedule_task_now(
+        stream->base.owning_connection->channel_slot->channel, &stream->cross_thread_work_task);
 }
 
-void aws_h1_stream_body_chunks_clean_up(struct aws_h1_stream *stream) {
-    AWS_PRECONDITION(stream);
-    if (!stream->body_chunks.lock.initialized) {
-        return;
+static void s_stream_cross_thread_work_task(struct aws_channel_task *task, void *arg, enum aws_task_status status) {
+    (void)task;
+    struct aws_h1_stream *stream = arg;
+    if (status != AWS_TASK_STATUS_RUN_READY) {
+        goto done;
     }
-    do {
-        if (NULL != stream->body_chunks.current_chunk) {
-            s_clean_up_body_chunk(&stream->body_chunks.current_chunk);
-        }
-    } while (aws_h1_populate_current_stream_chunk(&stream->body_chunks));
-    aws_mutex_clean_up(&stream->body_chunks.lock);
+
+    AWS_LOGF_TRACE(AWS_LS_HTTP_STREAM, "id=%p: Running stream cross-thread work task.", (void *)&stream->base);
+
+    /* BEGIN CRITICAL SECTION */
+    s_stream_lock_synced_data(stream);
```

Comment:
Maybe put an anonymous scope here like all the other critical sections?

Reply:
yeah, the anonymous scope is really just to make the block stand out more visually
we haven't been 100% consistent
will change where I can